---
title: "CSF Experiment"
output:
  html_document:
    code_folding: hide
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)
```

## 10X scRNA seq
All samples are initially processed using the 10X cellranger pipeline. This workflow aligns the reads to the reference genome and deconvolutes the cell barcodes to generate a cell x gene matrix. After a first pass of samples, multiple failed QC. Those samples were removed prior to final running of the secondary processing below. 

### Method snippet
If you wish to use a prescribed methods snippet:
*Raw reads were barcode deconvoluted, hash tag demultiplexed, and aligned to the reference genome (mm10) via cellranger (v6.0.1). All subsequent processing was performed in a dockerized environment using the [Seurat](https://www.cell.com/cell/fulltext/S0092-8674(21)00583-3?_returnURL=https%3A%2F%2Flinkinghub.elsevier.com%2Fretrieve%2Fpii%2FS0092867421005833%3Fshowall%3Dtrue) package (v4.1.1). Low quality cells (cells with percentage of reads of mitochondrial origin >10%, with percentage of reads of ribosomal origin >45%, with <500 features quantified, or with >3000 features quantified) were filtered from the dataset, and read counts were normalized using the [scTransform method](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-019-1874-1). Samples were integrated with the Seurat [integrate function](https://www.cell.com/cell/fulltext/S0092-8674(19)30559-8), clustered according to nearest neighbors using 30 principal components, and visualized via UMAP. Differential gene expression group cross-comparisons were performed via a pseudobulk analysis using [DESeq2](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-014-0550-8) (v1.36.0). Significant differentially expressed gene thresholds were defined at FDR adjusted p < 0.05.*

## Experimental set up

```{r Libraries,  message=FALSE}
#Basic manipulations
library(knitr)
library(tidyverse)
library(tibble)
library(dplyr)
library(hdf5r)
library(cowplot)
library(ggplot2)
library(ggalt)
library(ggpubr)
library(Matrix)
library(purrr)
library(reshape2)
library(RColorBrewer)
library(patchwork)
library(magrittr)
library(clustree)
library(Matrix.utils)

#scRNA core packages
library(Seurat)
library(S4Vectors)
library(scRNAseq)
library(SeuratWrappers)
library(SingleCellExperiment)
library(dittoSeq)
library(scDataviz)
library(DropletQC)
library(scCustomize)


#scRNA velocity, pseudotime, annotations, doublets
library(monocle3)
library(velocyto.R)
library(slingshot)
library(SingleR)
library(DoubletFinder)
library(SoupX)

#Other random packages for pseudobulk or things I like to use
library(apeglm)
library(scater)
library(magrittr)
library(pheatmap)
library(DESeq2)
library(edgeR)
library(EnhancedVolcano)
library(PCAtools)
library(clusterProfiler)

options(future.globals.maxSize=10000*1024^2)
```

## Filtering
The QC files can be found in the QC directory.In brief- cells are assessed for their overall transcript captures, percentage of mitochondrial content, and percentage of ribosomal content. Overall transcript depth can provide insight into low quality GEMs and putative multiplets. Percentage of ribosomal is less used, but can be used to evaluate cell presence and poorly captured GEMs. Percentage of mitochondrial is frequently used as a proxy for dead/dying cells. Since hashtags were used we did not implement any additional multiplet scrubbing. 

In this experiment, cells were removed if they had >4000 or <1000 nFeature_RNA, >10% mitochondrial reads, or >45% ribosomal reads. This was defined by the overall expression distribution and anticipated experimental parameters. 

```{r SamplesInput, warning=FALSE, message=FALSE, fig.align="center",results='hide'}
samplelist <- list();
dir.create('QC');
dir.create('clustering');
dir.create('cluster_markers');
dir.create('Gene_figures');
dir.create('pseudobulk');
dir.create('pseudobulk/counts');

cellcounts <- data.frame(matrix(ncol=2,nrow=0)); colnames(cellcounts) <- c('type','count')
postcellcounts <- data.frame(matrix(ncol=2,nrow=0)); colnames(postcellcounts) <- c('Sample','Count')
opt <- read.csv(file='metadata.csv')
for (i in 1:nrow(opt)){
   row <- opt[i,]
   filename <- row$samples
   samp <- row$name
   samp<-paste(samp)
   group <- row$subgroup
   group<-paste(group)
   
   batch <- row$batch
   batch<-paste(batch)
   
   directory <- row$directory
   
    data_in <- Read10X_h5(filename=paste(directory,filename,sep=''), use.names=TRUE, unique.features=TRUE);

    s <- CreateSeuratObject(counts=data_in, project=,min.cells=3, min.features=200);
    
    s@meta.data$samp <- paste(samp);
    s@meta.data$group <- paste(group);
    s@meta.data$batch <- paste(batch);
  
    s[['percent.mito']] <- PercentageFeatureSet(s, pattern='^MT-');
    s[['percent.ribo']] <- PercentageFeatureSet(s,pattern='^RP[SL][[:digit:]]');

    ## Visualize pre-filter
    jpeg(paste('./QC/',samp,'_pre_filt_geneplot_qc.jpg',sep=''))
    in_plot <- FeatureScatter(s, feature1='nCount_RNA',feature2='nFeature_RNA');
    par(mfrow=c(2,2))
    print(in_plot)
    dev.off()
    
    jpeg(paste('./QC/',samp,'_pre_filt_vln_qc.jpg',sep=''))
    in_plot <- VlnPlot(s, features=c('nFeature_RNA','nCount_RNA','percent.mito','percent.ribo'),ncol=4);
    print(in_plot)
    dev.off()

    cellcounts[nrow(cellcounts)+1,] <- list(sample=paste(samp, '_prefilt',sep=''), count=length(WhichCells(s)))
    
    ## Filter
    s <- subset (s, subset= nFeature_RNA < 4000 & percent.mito < 10 & percent.ribo < 45 & nFeature_RNA > 1000 );
    
    jpeg(paste('./QC/',samp,'_post_filt_vln_qc.jpg',sep=''))
    out_plot <- VlnPlot(s, features=c('nFeature_RNA','nCount_RNA','percent.mito','percent.ribo'),ncol=4);
    print(out_plot)
    dev.off()
    jpeg(paste('./QC/',samp,'_post_filt_geneplot_qc.jpg',sep=''))
    out_plot <- FeatureScatter(s, feature1='nCount_RNA',feature2='nFeature_RNA');
    print(out_plot)
    dev.off()
    
    cellcounts[nrow(cellcounts)+1,] <- list(sample=paste(samp, '_postfilt',sep=''), count=length(WhichCells(s)))
    postcellcounts[nrow(postcellcounts)+1,] <- list(Sample=paste(samp), Count=length(WhichCells(s)))
    
    ## Transform and regress the data
    #handled in the scrubbing step below
    #s <- SCTransform(s, vars.to.regress=c('percent.mito','percent.ribo'));
    #s <- RunPCA(s, verbose=F)

    ## Rename temp files
    assign(paste(samp),s);
    
    samplelist <- c(samplelist, eval(parse(text=samp)))
    
}
```

```{r DubScrub, warning=FALSE, message=FALSE, fig.align="center",results='hide'}
for (i in 1:length(samplelist)){
  samplelist[[i]] <- SCTransform(samplelist[[i]],vars.to.regress=c('percent.mito','percent.ribo'), verbose=FALSE)
  samplelist[[i]] <- RunPCA(samplelist[[i]])
  samplelist[[i]] <- RunUMAP(samplelist[[i]], dims=1:10)
  samplelist[[i]] <- FindNeighbors(samplelist[[i]], reduction='pca', dims=1:10); 
  samplelist[[i]] <- FindClusters(samplelist[[i]],resolution=0.5); 
  
  sweep.res.list <- paramSweep_v3(samplelist[[i]], PCs=1:10, sct=TRUE)
  sweep.stats <- summarizeSweep(sweep.res.list, GT=FALSE)
  bcmvn <- find.pK(sweep.stats)
  
  #homotypic.prop <- modelHomotypic(seurat_clusters)
  nExp_poi <- round(0.1*nrow(samplelist[[i]]@meta.data))
  #nExp_poi.adj <- round(nExp_poi*(1-homotypic.prop))
  
  samplelist[[i]] <- doubletFinder_v3(samplelist[[i]], PCs=1:10, pN=0.25, pK=0.09, nExp=nExp_poi, reuse.pANN=FALSE, sct=TRUE)
  #samplelist[[i]] <- doubletFinder_v3(samplelist[[i]], PCs=1:10, pN=0.25, pK=0.09, nExp=nExp_poi.adj, reuse.pANN = "pANN_0.25_0.09_913", sct=TRUE)
  
}

for(i in 1:length(samplelist)){
  DF.name = colnames(samplelist[[i]]@meta.data)[grepl("DF.classification", colnames(samplelist[[i]]@meta.data))]
  samplelist[[i]] = samplelist[[i]][, samplelist[[i]]@meta.data[, DF.name] == "Singlet"]
  
  cellcounts[nrow(cellcounts)+1,] <- list(sample=paste(unique(samplelist[[i]]$samp), '_post_scrub',sep=''), count=length(WhichCells(samplelist[[i]])))
}
write.table(cellcounts, './QC/cellcounts_prepost_filter.txt',sep='\t')
```

```{r IntegrateSamples, warning=FALSE, message=FALSE, fig.align="center",results='hide'}
#First identify the features we can use for integration. You can use all the features, but biologically it doesnt change much. Most of this is carried in the first 2000 features. We then identify the agnostic anchor points, and then perform the integration
combined.features <- SelectIntegrationFeatures(object.list=samplelist, nfeatures=2000);
combined.list <- PrepSCTIntegration(object.list=samplelist, anchor.features=combined.features, verbose=FALSE);
#The reference are all the NHC samples
combined.anchors <- FindIntegrationAnchors(object.list=combined.list, normalization.method='SCT', anchor.features=combined.features, reference=c(3,13,14,16,17,18,20,21,22,24,25,26),verbose=FALSE, reduction='rpca');
combined <- IntegrateData(anchorset=combined.anchors, normalization.method='SCT')
```

The final filtered distribution is shown here. 
```{r PostFilt, warning=FALSE, message=FALSE, fig.width=10, fig.height=10, fig.align="center",results='hide'}
VlnPlot(combined, features=c('nFeature_RNA','nCount_RNA','percent.mito','percent.ribo'),pt.size=0,ncol=4)
```

```{r CountTable, warning=FALSE, message=FALSE, fig.width=10, fig.height=10, fig.align="center",results='hide'}
kable(postcellcounts)
```

## Integration and clustering
After filtering and transformation, the samples are integrated. This is to align cell types across the samples and reduce batch effects. Aligning the cell types across samples is essential for later group cross comparisons, and ensure false positive "unique" clusters are not generated based off of batch-to-batch differences. If a population of cells is truly unique to a sample or group, it will still cluster as an isolated population.

Clustering is performed after the integration. Cells that are similar as determined by a weighted nearest neighbor approach are grouped into "clusters" based on their transcriptional profiles. This is then visualized via UMAP. This clustering resolution can be tweaked to either aggregate or be further broken down. I personally am a fan of starting in the middle ground- clusters that are biologically similar can then be manually aggregated, or specific populations can be isolated and reclustered irrespective of other populations to dive deeper into subpopulations. 
```{r Clustering, warning=FALSE, message=FALSE, fig.width=10, fig.height=10, fig.align="center",results='hide'}
#Followed by the PCA, UMAP, and neighbor identification. 
combined <- RunPCA(combined, verbose=FALSE, npcs=30);
combined <- FindNeighbors(combined, reduction='pca', dims=1:10); 
combined <- RunUMAP(combined, reduction='pca', dims=1:10); 
combined <- FindClusters(combined, resolution=0.5); 

write.csv(table(Idents(combined),combined$samp), file='./clustering/counts_per_cluster.txt')

p1 <- DimPlot(combined, reduction='umap', label=TRUE, pt.size=2, label.size=10); 
jpeg('./clustering/umap_solo.jpg',height=800, width=1500)
print(p1)
dev.off()

jpeg('./clustering/umap_group.jpg',height=800,width=1500)
DimPlot(combined, reduction='umap',split.by='group')
dev.off()
jpeg('./clustering/umap_samples.jpg',height=800, width=1500)
DimPlot(combined, reduction='umap',split.by='samp');
dev.off()

DefaultAssay(combined) <- 'SCT'
p1
```

### Identifying cluster markers
Cluster markers can be defined by comparing the transcriptional profile of the respective cluster vs all other cells.

```{r Markers, warning=FALSE, message=FALSE, fig.align="center"}
DefaultAssay(combined) <- 'SCT'
combined <- PrepSCTFindMarkers(combined)
clustnum <- nrow(table(Idents(combined),combined$samp))
for (i in 0:(clustnum-1)){
  clust.markers <- FindMarkers(combined, ident.1=i, print.bar=FALSE)
  write.csv(clust.markers, paste('./cluster_markers/cluster_',i,'.csv', sep=''))
  assign(paste('clust', i,'.markers', sep=''), clust.markers)
}

#Or we can use the FindAllMarkers commmand to find all markers, not just one cluster at a time
markers <- FindAllMarkers(combined, only.pos=TRUE, min.pct=0.25, logfc.threshold=0.25)

#And we can print out real quick the top 3 per cluster
kable(markers %>% group_by(cluster) %>% slice_max(n=3, order_by=avg_log2FC))
```
```{r table1, echo=FALSE, message=FALSE, warnings=FALSE, results='asis'}
tabl <- "
| Gene | p_val | avg_log2FC | pct.1 | pct.2 | p_val_adj | cluster |
| ---- | ----- | ---------- | ----- | ----- | --------- | ------- |
| Gene name | Raw p value | Average fold change for this cluster vs all other clusters in log2 scale | Percentage of cells in this cluster expressing the gene | Percentage of ALL other cells expressing the gene | FDR corrected p value | Which respective cluster this is for | 

"
cat(tabl)
```
As shown in the table, some genes can have large or significant changes but still be very minimally expressed. It usually requires a customized strategy to define populations. When working with macro populations- IE all T cells vs all B cells- there are very strong markers that can delineate the two that should have minimal cross expression. When diving into subtypes- such as effector CD8 T cells vs memory CD8 T cells, the differences can become more "shades of grey" and difficult to delineate without deep domain expertise. 

### Automated cluster identification
There are also a number of automated cell annotations tools. I use these with a grain of salt- it is highly predicated that a highly similar cell population has been previously sequenced and annotated correctly. When getting into extreme subpopulations these tools are often next-to-useless, but for macro clusters can be highly helpful. 

```{r AutoMarkers, warning=FALSE, message=FALSE, fig.width=10, fig.height=10, fig.align="center",results='hide'}
Idents(combined) <- 'seurat_clusters'
library(SingleR)
library(scRNAseq)
se <- HumanPrimaryCellAtlasData() #General human
test <- SingleR(test=as.SingleCellExperiment(combined, assay='RNA'), ref=se, assay.type.test=1, labels=se$label.main, clusters=combined$seurat_clusters)

new.cluster.ids <- c(test$labels) #Pull from clusters of singleR
names(new.cluster.ids) <- levels(combined)
combined <- RenameIdents(combined, new.cluster.ids)

GenHuman_main <- DimPlot(combined, reduction='umap', label=TRUE, label.size=6, pt.size=1.5)+ NoLegend()

jpeg('./clustering/umap_labelled_singleRclusts_generalhuman.jpg',height=800, width=1500)
DimPlot(combined, reduction='umap', label=TRUE, label.size=6, pt.size=1.5)+ NoLegend()
dev.off()
#################
Idents(combined) <- 'seurat_clusters'
se <- MonacoImmuneData() 
test <- SingleR(test=as.SingleCellExperiment(combined, assay='RNA'), ref=se, assay.type.test=1, labels=se$label.main, clusters=combined$seurat_clusters)

new.cluster.ids <- c(test$labels) #Pull from clusters of singleR
names(new.cluster.ids) <- levels(combined)
combined <- RenameIdents(combined, new.cluster.ids)

Moanco_main <- DimPlot(combined, reduction='umap', label=TRUE, label.size=6, pt.size=1.5)+ NoLegend()

jpeg('./clustering/umap_labelled_singleRclusts_monaco.jpg',height=800, width=1500)
DimPlot(combined, reduction='umap', label=TRUE, label.size=6, pt.size=1.5)+ NoLegend()
dev.off()

GenHuman_main
```

```{r Annotations, warning=FALSE, message=FALSE, fig.align="center",results='hide'}
combined <- data$FullClustering
DefaultAssay(combined) <- 'SCT'
Idents(combined) <- 'seurat_clusters'
new.cluster.ids <- c('T cells','T cells','Monocytes','T cells','T cells','T cells','T cells','Monocytes','Monocytes','Monocytes','T cells','T cells','T cells','Monocytes','B cells','Monocytes','T cells')
names(new.cluster.ids) <- levels(combined)
combined <- RenameIdents(combined, new.cluster.ids)
combined[['Annotations']] <- Idents(object=combined)
write.csv(table(Idents(combined), combined$samp), file='base_counts.csv')

jpeg('./clustering/UMAP_manual_annotation.jpg',height=800, width=1500)
DimPlot(combined, reduction='umap', label=TRUE, label.size=6, pt.size=1.5)+ NoLegend()
dev.off()

opt <- read.csv(file='metadata.csv')

DimPlot(combined, reduction='umap', label=TRUE, label.size=6, pt.size=1.5)+ NoLegend()

jpeg('./clustering/VlnPlot_markers.jpg',height=1200, width=800)
VlnPlot(combined, features=c('MS4A1','CD3E','CD8A','CD4','FOXP3','HBB','SLC25A37','PPBP','GNLY','XCL1','XCL2','NKG7','CD14','FCGR3A','CST3','FCER1A','BCL11A','JCHAIN','ITM2C','IGKC'), stack=TRUE, sort=TRUE, flip=TRUE) + theme(legend.position='none')
dev.off()

```

## Pseudobulk Analysis
Thanks to the replicates provided by hashtags, we can perform psuedobulk analysis. This has been shown to be more accurate for differential gene expression comparisons, with less p-inflation that can occur. At this point, the data can be visualized akin to bulk RNA-seq data on a cluster basis across groups. 
```{r PseudobulkPrep, warning=FALSE, message=FALSE, fig.align="center",results='hide'}
combined$samp <- gsub("_","", as.character(combined$samp))
Idents(combined) <- 'samp'
combined[['sample_id']] <- Idents(combined)
Idents(combined) <- 'Annotations'

Idents(combined) <- 'group'
combined[['group_id']] <- Idents(combined)
Idents(combined) <- 'Annotations'

counts <- combined@assays$RNA@counts
metadata <- combined@meta.data
metadata$cluster_id <- factor(combined@active.ident)
sce <- SingleCellExperiment(assays=list(counts=counts), colData=metadata)
groups <- colData(sce, c("cluster_id", "sample_id"))

kids <- purrr::set_names(levels(sce$cluster_id))

nk <- length(kids)

sids <- purrr::set_names(levels(sce$sample_id))

ns <- length(sids)

n_cells <- as.numeric(table(sce$sample_id))

m <- match(sids, sce$sample_id)

ei <- data.frame(colData(sce)[m, ], 
    n_cells, row.names = NULL) %>% 
    select(-"cluster_id")

groups <- colData(sce)[, c("cluster_id", "sample_id")]

pb <- aggregate.Matrix(t(counts(sce)), 
    groupings = groups, fun = "sum") 

splitf <- sapply(stringr::str_split(rownames(pb), 
                                    pattern = "_",  
                                    n = 2), 
                 `[`, 1)

pb <- split.data.frame(pb, 
                       factor(splitf)) %>%
        lapply(function(u) 
                set_colnames(t(u), 
                             stringr::str_extract(rownames(u), "(?<=_)[:alnum:]+")))

get_sample_ids <- function(x){
        pb[[x]] %>%
                colnames()
}

de_samples <- map(1:length(kids), get_sample_ids) %>%
        unlist()

samples_list <- map(1:length(kids), get_sample_ids)

get_cluster_ids <- function(x){
        rep(names(pb)[x], 
            each = length(samples_list[[x]]))
}

de_cluster_ids <- map(1:length(kids), get_cluster_ids) %>%
        unlist()


gg_df <- data.frame(cluster_id = de_cluster_ids,
                    sample_id = de_samples)

gg_df <- left_join(gg_df, ei[, c("sample_id", "group_id")]) 


metadata <- gg_df %>%
        dplyr::select(cluster_id, sample_id, group_id) 
```

### DESeq Table headers:
```{r table2, echo=FALSE, message=FALSE, warnings=FALSE, results='asis'}
tabl <- "
| Gene | Geneid | baseMean | log2FoldChange | lfcSE | stat | pvalue | padj |
| ---- | ------ | -------- | -------------- | ----- | ---- | ------ | ---- |
| Ensembl gene code | Gene symbol | Average of normalized count values over all samples. A good check for low count genes | Fold change between the compared group in log2 scale. +1 = doubled, -1 = halved in comparison. | Standard error value | Wald statistic | Raw pvalue | FDR correction pvalue |

"
cat(tabl)
```

```{r Pseudobulk2, warning=FALSE, message=FALSE, fig.align="center",results='hide'}
direct_compares <- list()
mats <- list()
clusters <- unique(metadata$cluster_id)
for(i in clusters){
    cluster_metadata <- metadata[which(metadata$cluster_id == paste(i)), ]
    rownames(cluster_metadata) <- cluster_metadata$sample_id
    counts <- pb[[paste(i)]]
    cluster_counts <- data.frame(counts[, which(colnames(counts) %in% rownames(cluster_metadata))])
    all(rownames(cluster_metadata) == colnames(cluster_counts))    
    dds <- DESeqDataSetFromMatrix(cluster_counts, 
                              colData = cluster_metadata, 
                              design = ~ group_id)
    dds <- DESeq(dds)
    vst <- DESeq2::varianceStabilizingTransformation(dds, blind=TRUE)
    mat <- assay(vst)

    compares <- list(
      list('CovidNormalTesting','control'),
      list('CovidPreRecovery','control'),
      list('CovidPersistent','control')
      
    )
    
    for(co in compares){
      run <- try({
        contrast <- c("group_id", co[[1]], co[[2]])
        res <- results(dds, 
            contrast = contrast,
            alpha = 0.05)
    
        res <- lfcShrink(dds, 
            contrast =  contrast,
            res=res,type="normal") 
        
        res <- res[order(res$padj),]
    
        res_tbl <- res %>%
            data.frame() %>%
            rownames_to_column(var="gene") %>%
            as_tibble()
        
        assign(paste0("cluster", i, "_", co[[1]], "_vs_", co[[2]]), res_tbl)
        jpeg(paste0("./pseudobulk/","volcano_", i, "_", co[[1]], "_vs_", co[[2]], ".jpg"), height=800, width=1000)
        p <- EnhancedVolcano(res_tbl,
                             lab=as.character(res_tbl$gene),
                             title=paste0(i, " ", co[[1]], " vs ", co[[2]], ".csv"),
                             x='log2FoldChange',
                             y='padj',
                             legendPosition = 'bottom',
                              legendLabels=c('Not sig.','Log (base 2) FC','padj-value','padj-value & Log (base 2) FC'),
                              pCutoff=0.05, 
                              FCcutoff=1
                             )
        print(p)
        dev.off()
        write.csv(res_tbl,
            paste0("./pseudobulk/","cluster", i, "_", co[[1]], "_vs_", co[[2]], "_all_genes.csv"),
            quote = FALSE, 
            row.names = FALSE)
        write.csv(as.data.frame(mat), file=paste0("./pseudobulk/counts/","cluster", i, "_vst_counts.csv"))  
        write.csv(counts(dds,normalized = TRUE), file=paste0("./pseudobulk/counts/", "cluster",i, "_dds_counts.csv")) 
        
        direct_compares[[paste0("cluster", i, "_", co[[1]], "_vs_", co[[2]])]] <- res_tbl
        mats[[paste0("cluster", i, "_", co[[1]], "_vs_", co[[2]])]] <- as.data.frame(mat)
      })
      if(inherits(run, 'try-error')) {
        next
      }
    }
}
```

## Population shifts
```{r Popshifts, warning=FALSE, message=FALSE, fig.align="center",results='hide'}
dir.create('PopShifts');
Idents(combined) <- 'seurat_clusters'
counts <- table(Idents(combined), combined$samp)
counts <- prop.table(counts, margin=2)*100

sigcounts <- data.frame(matrix(ncol=2,nrow=0)); colnames(sigcounts) <- c('cluster','pval')

temp <- as.data.frame(t(counts))
names(temp) <- c('sample_id','cluster_id','Freq')
temp$sample_id <- gsub("_","", as.character(temp$sample_id))

temp <-merge(temp, metadata %>% select(sample_id, group_id), by='sample_id') %>% distinct(sample_id, cluster_id,.keep_all=TRUE)

for(i in unique(temp$cluster_id)){
  test <- aov(Freq ~ group_id, data=subset(temp, cluster_id==i))
  sigcounts[nrow(sigcounts)+1,]<-list(cluster=paste(i), pval=paste(summary(test)[[1]][[5]][[1]]))
}

write.csv(sigcounts, './PopShifts/anova_cellpercent.csv')

p <- ggplot(temp, aes(x=group_id, y=Freq, fill=cluster_id)) + geom_bar(position='fill', stat='identity') + theme_classic()
p <- p + labs(y='Percentage of cells', x='Sample')
p <- p + scale_fill_discrete(name = "Cluster") + theme(axis.text.x=element_text(angle=270))

jpeg('./PopShifts/cluster_ratios.jpg',height=800, width=1500)
p
dev.off()
p


Idents(combined) <- 'Annotations'
countsA <- table(Idents(combined), combined$samp)
countsA <- prop.table(countsA, margin=2)*100

sigcounts <- data.frame(matrix(ncol=2,nrow=0)); colnames(sigcounts) <- c('cluster','pval')

tempA <- as.data.frame(t(countsA))
names(tempA) <- c('sample_id','cluster_id','Freq')
tempA$sample_id <- gsub("_","", as.character(tempA$sample_id))

tempA <-merge(tempA, metadata %>% select(sample_id, group_id), by='sample_id') %>% distinct(sample_id, cluster_id,.keep_all=TRUE)

for(i in unique(tempA$cluster_id)){
  test <- aov(Freq ~ group_id, data=subset(tempA, cluster_id==i))
  sigcounts[nrow(sigcounts)+1,]<-list(cluster=paste(i), pval=paste(summary(test)[[1]][[5]][[1]]))
}

write.csv(sigcounts, './PopShifts/anova_cellpercent_annotations.csv')

p <- ggplot(tempA, aes(x=group_id, y=Freq, fill=cluster_id)) + geom_bar(position='fill', stat='identity') + theme_classic()
p <- p + labs(y='Percentage of cells', x='Sample')
p <- p + scale_fill_discrete(name = "Cluster") + theme(axis.text.x=element_text(angle=270))

jpeg('./PopShifts/annotation_ratios.jpg',height=800, width=1500)
p
dev.off()
p

```

### PCA plot of proportion table
```{r PCA, warning=FALSE, message=FALSE, fig.align="center",results='hide'}
for(i in unique(tempA$cluster_id)){
  jpeg(paste0('./PopShifts//cell_proportion_boxplot_',i,'.jpg'),height=800, width=800)
  p <- ggplot(tempA[tempA$cluster_id==paste(i),], aes(x=group_id, y=Freq)) + geom_boxplot(fill='#A4A4A4', color="black")+theme_classic() + labs(y="Cell proportion per sample (%)", x="Group") + theme(axis.text.x = element_text(angle = 45, hjust=1), text=element_text(size=20))
  print(p)
  dev.off()
}

countstemp <- counts[,order(colnames(counts))]
metatemp <- opt %>% column_to_rownames('name')
metatemp <- metatemp[order(row.names(metatemp)),]

p <- pca(countstemp, metadata=metatemp)
plot <- biplot(p, colby='subgroup',
               encircle=TRUE,
            legendPosition='top', legendLabSize=16, legendIconSize=8.0,
            pointSize=6,
            labSize=0)
jpeg('./pca_cell_props.jpg', height=1200, width=1200)
print(plot)
dev.off()

plotL <- biplot(p, colby='subgroup',
               encircle=TRUE,
               showLoadings=TRUE,
               sizeLoadingsNames = 6,
               ntopLoadings=8,
            legendPosition='top', legendLabSize=16, legendIconSize=8.0,
            pointSize=6,
            labSize=0)
jpeg('./pca_cell_props_loadings.jpg', height=1200, width=1200)
print(plotL)
dev.off()

jpeg('./loadings.jpg',height=1200, width=1200)
plotloadings(p, labSize=6, rangeRetain=0.5)
dev.off()

print(plot)
```

# Monocyte reclustering

```{r MonoRecluster, warning=FALSE, message=FALSE, fig.align="center",results='hide'}
Idents(combined) <- 'seurat_clusters'
monos <- subset(combined, idents=c(2,8,7,9,15,13))
DefaultAssay(monos) <- 'RNA'

monos <- DietSeurat(monos, assays='RNA')
monos.list <- SplitObject(monos, split.by='samp') #split by sample makes monster object???
for (i in 1:length(monos.list)){
  monos.list[[i]] <- SCTransform(monos.list[[i]], vars.to.regress=c('percent.mito','percent.ribo'),verbose=FALSE)
}

options(future.globals.maxSize=1000000*1024^2)

monos.features <- SelectIntegrationFeatures(object.list=monos.list, nfeatures=2000);
monos.list <- PrepSCTIntegration(object.list=monos.list, anchor.features=monos.features, verbose=FALSE);
monos.anchors <- FindIntegrationAnchors(object.list=monos.list, normalization.method='SCT', anchor.features=monos.features, verbose=FALSE);
monos <- IntegrateData(anchorset=monos.anchors, normalization.method='SCT', k.weight=25);

monos <- RunPCA(monos, verbose=FALSE, npcs=30);
monos <- RunUMAP(monos, reduction='pca', dims=1:10); 
monos <- FindNeighbors(monos, reduction='pca', dims=1:10); 
monos <- FindClusters(monos,resolution=0.2); 

DefaultAssay(monos) <- 'SCT'

dir.create('./monos/')
dir.create('./monos/clustering/')
p1 <- DimPlot(monos, reduction='umap', label=TRUE, pt.size=2, label.size=10); 
jpeg('./monos/clustering/umap_solo.jpg',height=800, width=1500)
print(p1)
dev.off()

jpeg('./monos/clustering/umap_group.jpg',height=800,width=1500)
DimPlot(monos, reduction='umap',split.by='group')
dev.off()
jpeg('./monos/clustering/umap_samples.jpg',height=800, width=1500)
DimPlot(monos, reduction='umap',split.by='samp');
dev.off()

dir.create('./monos/cluster_markers/')
monos <- PrepSCTFindMarkers(monos)
clustnum <- nrow(table(Idents(monos),monos$samp))
for (i in 0:(clustnum-1)){
  clust.markers <- FindMarkers(monos, ident.1=i, print.bar=FALSE)
  write.csv(clust.markers, paste('./monos/cluster_markers/cluster_',i,'.csv', sep=''))
  assign(paste('mac_clust', i,'.markers', sep=''), clust.markers)
}

```

## Monocyte pseudobulk 
```{r MonoPseudobulkPrep, warning=FALSE, message=FALSE, fig.align="center",results='hide'}
direct_compares <- list()
mats <- list()


monos$samp <- gsub("_","", as.character(monos$samp))
Idents(monos) <- 'samp'
monos[['sample_id']] <- Idents(monos)
Idents(monos) <- 'seurat_clusters'

Idents(monos) <- 'group'
monos[['group_id']] <- Idents(monos)
Idents(monos) <- 'seurat_clusters'

counts <- monos@assays$RNA@counts
metadata <- monos@meta.data
metadata$cluster_id <- factor(monos@active.ident)
sce <- SingleCellExperiment(assays=list(counts=counts), colData=metadata)
groups <- colData(sce, c("cluster_id", "sample_id"))

kids <- purrr::set_names(levels(sce$cluster_id))

nk <- length(kids)

sids <- purrr::set_names(levels(sce$sample_id))

ns <- length(sids)

n_cells <- as.numeric(table(sce$sample_id))

m <- match(sids, sce$sample_id)

ei <- data.frame(colData(sce)[m, ], 
    n_cells, row.names = NULL) %>% 
    select(-"cluster_id")

groups <- colData(sce)[, c("cluster_id", "sample_id")]

pb <- aggregate.Matrix(t(counts(sce)), 
    groupings = groups, fun = "sum") 

splitf <- sapply(stringr::str_split(rownames(pb), 
                                    pattern = "_",  
                                    n = 2), 
                 `[`, 1)

pb <- split.data.frame(pb, 
                       factor(splitf)) %>%
        lapply(function(u) 
                set_colnames(t(u), 
                             stringr::str_extract(rownames(u), "(?<=_)[:alnum:]+")))

get_sample_ids <- function(x){
        pb[[x]] %>%
                colnames()
}

de_samples <- map(1:length(kids), get_sample_ids) %>%
        unlist()

samples_list <- map(1:length(kids), get_sample_ids)

get_cluster_ids <- function(x){
        rep(names(pb)[x], 
            each = length(samples_list[[x]]))
}

de_cluster_ids <- map(1:length(kids), get_cluster_ids) %>%
        unlist()


gg_df <- data.frame(cluster_id = de_cluster_ids,
                    sample_id = de_samples)

gg_df <- left_join(gg_df, ei[, c("sample_id", "group_id")]) 


metadata <- gg_df %>%
        dplyr::select(cluster_id, sample_id, group_id) 

clusters <- unique(metadata$cluster_id)
for(i in clusters){
    cluster_metadata <- metadata[which(metadata$cluster_id == paste(i)), ]
    rownames(cluster_metadata) <- cluster_metadata$sample_id
    counts <- pb[[paste(i)]]
    cluster_counts <- data.frame(counts[, which(colnames(counts) %in% rownames(cluster_metadata))])
    all(rownames(cluster_metadata) == colnames(cluster_counts))    
    dds <- DESeqDataSetFromMatrix(cluster_counts, 
                              colData = cluster_metadata, 
                              design = ~ group_id)
    dds <- DESeq(dds)
    vst <- DESeq2::varianceStabilizingTransformation(dds, blind=TRUE)
    mat <- assay(vst)

    compares <- list(
      list('CovidNormalTesting','control'),
      list('CovidPreRecovery','control'),
      list('CovidPersistent','control')
      
    )
    
    for(co in compares){
      run <- try({
        contrast <- c("group_id", co[[1]], co[[2]])
        res <- results(dds, 
            contrast = contrast,
            alpha = 0.05)
    
        res <- lfcShrink(dds, 
            contrast =  contrast,
            res=res,type="normal") 
        
        res <- res[order(res$padj),]
    
        res_tbl <- res %>%
            data.frame() %>%
            rownames_to_column(var="gene") %>%
            as_tibble()
        
        assign(paste0("cluster", i, "_", co[[1]], "_vs_", co[[2]]), res_tbl)
        jpeg(paste0("./pseudobulk/monos/","volcano_", i, "_", co[[1]], "_vs_", co[[2]], ".jpg"), height=800, width=1000)
        p <- EnhancedVolcano(res_tbl,
                             lab=as.character(res_tbl$gene),
                             title=paste0(i, " ", co[[1]], " vs ", co[[2]], ".csv"),
                             x='log2FoldChange',
                             y='padj',
                             legendPosition = 'bottom',
                              legendLabels=c('Not sig.','Log (base 2) FC','padj-value','padj-value & Log (base 2) FC'),
                              pCutoff=0.05, 
                              FCcutoff=1
                             )
        print(p)
        dev.off()
        write.csv(res_tbl,
            paste0("./pseudobulk/monos/","cluster", i, "_", co[[1]], "_vs_", co[[2]], "_all_genes.csv"),
            quote = FALSE, 
            row.names = FALSE)
        write.csv(as.data.frame(mat), file=paste0("./pseudobulk/monos/counts/","cluster", i, "_vst_counts.csv"))  
        write.csv(counts(dds,normalized = TRUE), file=paste0("./pseudobulk/monos/counts/", "cluster",i, "_dds_counts.csv")) 
        
        direct_compares[[paste0("monocyte_cluster", i, "_", co[[1]], "_vs_", co[[2]])]] <- res_tbl
        mats[[paste0("monocyte_cluster", i, "_", co[[1]], "_vs_", co[[2]])]] <- as.data.frame(mat)
      })
      if(inherits(run, 'try-error')) {
        next
      }
    }
}
```

## Tcell reclustering

```{r TCellRecluster, warning=FALSE, message=FALSE, fig.align="center",results='hide'}
Idents(combined) <- 'seurat_clusters'
tcells <- subset(combined, idents=c(0,1,3,16,10,6,4,5,12,11))
DefaultAssay(tcells) <- 'RNA'

tcells <- DietSeurat(tcells, assays='RNA')
tcells.list <- SplitObject(tcells, split.by='samp') #split by sample makes monster object???
for (i in 1:length(tcells.list)){
  tcells.list[[i]] <- SCTransform(tcells.list[[i]], vars.to.regress=c('percent.mito','percent.ribo'),verbose=FALSE)
  tcells.list[[i]] <- RunPCA(tcells.list[[i]], verbose=F)
}


tcells.features <- SelectIntegrationFeatures(object.list=tcells.list, nfeatures=2000);
tcells.list <- PrepSCTIntegration(object.list=tcells.list, anchor.features=tcells.features, verbose=FALSE);
tcells.anchors <- FindIntegrationAnchors(object.list=tcells.list, normalization.method='SCT', reference=c(3,13,14,16,17,18,20,21,22,24,25,26), anchor.features=tcells.features, verbose=FALSE);
tcells <- IntegrateData(anchorset=tcells.anchors, normalization.method='SCT');

tcells <- RunPCA(tcells, verbose=FALSE, npcs=30);
tcells <- RunUMAP(tcells, reduction='pca', dims=1:10); 
tcells <- FindNeighbors(tcells, reduction='pca', dims=1:10); 
tcells <- FindClusters(tcells,resolution=0.3); 

DefaultAssay(tcells) <- 'SCT'

dir.create('./tcells/')
dir.create('./tcells/clustering/')
p1 <- DimPlot(tcells, reduction='umap', label=TRUE, pt.size=2, label.size=10); 
jpeg('./tcells/clustering/umap_solo.jpg',height=800, width=1500)
print(p1)
dev.off()

jpeg('./tcells/clustering/umap_group.jpg',height=800,width=1500)
DimPlot(tcells, reduction='umap',split.by='group')
dev.off()
jpeg('./tcells/clustering/umap_samples.jpg',height=800, width=1500)
DimPlot(tcells, reduction='umap',split.by='samp');
dev.off()

dir.create('./tcells/cluster_markers/')
tcells <- PrepSCTFindMarkers(tcells)
clustnum <- nrow(table(Idents(tcells),tcells$samp))
for (i in 0:(clustnum-1)){
  clust.markers <- FindMarkers(tcells, ident.1=i, print.bar=FALSE)
  write.csv(clust.markers, paste('./tcells/cluster_markers/cluster_',i,'.csv', sep=''))
  assign(paste('mac_clust', i,'.markers', sep=''), clust.markers)
}

```

```{r TCellPseudobulkPrep, warning=FALSE, message=FALSE, fig.align="center",results='hide'}
tcells$samp <- gsub("_","", as.character(tcells$samp))
Idents(tcells) <- 'samp'
tcells[['sample_id']] <- Idents(tcells)
Idents(tcells) <- 'seurat_clusters'

Idents(tcells) <- 'group'
tcells[['group_id']] <- Idents(tcells)
Idents(tcells) <- 'seurat_clusters'

counts <- tcells@assays$RNA@counts
metadata <- tcells@meta.data
metadata$cluster_id <- factor(tcells@active.ident)
sce <- SingleCellExperiment(assays=list(counts=counts), colData=metadata)
groups <- colData(sce, c("cluster_id", "sample_id"))

kids <- purrr::set_names(levels(sce$cluster_id))

nk <- length(kids)

sids <- purrr::set_names(levels(sce$sample_id))

ns <- length(sids)

n_cells <- as.numeric(table(sce$sample_id))

m <- match(sids, sce$sample_id)

ei <- data.frame(colData(sce)[m, ], 
    n_cells, row.names = NULL) %>% 
    select(-"cluster_id")

groups <- colData(sce)[, c("cluster_id", "sample_id")]

pb <- aggregate.Matrix(t(counts(sce)), 
    groupings = groups, fun = "sum") 

splitf <- sapply(stringr::str_split(rownames(pb), 
                                    pattern = "_",  
                                    n = 2), 
                 `[`, 1)

pb <- split.data.frame(pb, 
                       factor(splitf)) %>%
        lapply(function(u) 
                set_colnames(t(u), 
                             stringr::str_extract(rownames(u), "(?<=_)[:alnum:]+")))

get_sample_ids <- function(x){
        pb[[x]] %>%
                colnames()
}

de_samples <- map(1:length(kids), get_sample_ids) %>%
        unlist()

samples_list <- map(1:length(kids), get_sample_ids)

get_cluster_ids <- function(x){
        rep(names(pb)[x], 
            each = length(samples_list[[x]]))
}

de_cluster_ids <- map(1:length(kids), get_cluster_ids) %>%
        unlist()


gg_df <- data.frame(cluster_id = de_cluster_ids,
                    sample_id = de_samples)

gg_df <- left_join(gg_df, ei[, c("sample_id", "group_id")]) 


metadata <- gg_df %>%
        dplyr::select(cluster_id, sample_id, group_id) 

clusters <- unique(metadata$cluster_id)
for(i in clusters){
    cluster_metadata <- metadata[which(metadata$cluster_id == paste(i)), ]
    rownames(cluster_metadata) <- cluster_metadata$sample_id
    counts <- pb[[paste(i)]]
    cluster_counts <- data.frame(counts[, which(colnames(counts) %in% rownames(cluster_metadata))])
    all(rownames(cluster_metadata) == colnames(cluster_counts))    
    dds <- DESeqDataSetFromMatrix(cluster_counts, 
                              colData = cluster_metadata, 
                              design = ~ group_id)
    dds <- DESeq(dds)
    vst <- DESeq2::varianceStabilizingTransformation(dds, blind=TRUE)
    mat <- assay(vst)

    compares <- list(
      list('CovidNormalTesting','control'),
      list('CovidPreRecovery','control'),
      list('CovidPersistent','control')
      
    )
    
    for(co in compares){
      run <- try({
        contrast <- c("group_id", co[[1]], co[[2]])
        res <- results(dds, 
            contrast = contrast,
            alpha = 0.05)
    
        res <- lfcShrink(dds, 
            contrast =  contrast,
            res=res,type="normal") 
        
        res <- res[order(res$padj),]
    
        res_tbl <- res %>%
            data.frame() %>%
            rownames_to_column(var="gene") %>%
            as_tibble()
        
        assign(paste0("cluster", i, "_", co[[1]], "_vs_", co[[2]]), res_tbl)
        jpeg(paste0("./pseudobulk/tcells/","volcano_", i, "_", co[[1]], "_vs_", co[[2]], ".jpg"), height=800, width=1000)
        p <- EnhancedVolcano(res_tbl,
                             lab=as.character(res_tbl$gene),
                             title=paste0(i, " ", co[[1]], " vs ", co[[2]], ".csv"),
                             x='log2FoldChange',
                             y='padj',
                             legendPosition = 'bottom',
                              legendLabels=c('Not sig.','Log (base 2) FC','padj-value','padj-value & Log (base 2) FC'),
                              pCutoff=0.05, 
                              FCcutoff=1
                             )
        print(p)
        dev.off()
        write.csv(res_tbl,
            paste0("./pseudobulk/tcells/","cluster", i, "_", co[[1]], "_vs_", co[[2]], "_all_genes.csv"),
            quote = FALSE, 
            row.names = FALSE)
        write.csv(as.data.frame(mat), file=paste0("./pseudobulk/tcells/counts/","cluster", i, "_vst_counts.csv"))  
        write.csv(counts(dds,normalized = TRUE), file=paste0("./pseudobulk/tcells/counts/", "cluster",i, "_dds_counts.csv")) 
        
        direct_compares[[paste0("tcell_cluster", i, "_", co[[1]], "_vs_", co[[2]])]] <- res_tbl
        mats[[paste0("tcell_cluster", i, "_", co[[1]], "_vs_", co[[2]])]] <- as.data.frame(mat)
      })
      if(inherits(run, 'try-error')) {
        next
      }
    }
}
```

## Annotation follow ups

```{r SubClusterAnnotation, warning=FALSE, message=FALSE, fig.align="center",results='hide'}
Idents(tcells) <- 'seurat_clusters'
se <- celldex::MonacoImmuneData() 
test <- SingleR(test=as.SingleCellExperiment(tcells, assay='RNA'), ref=se, assay.type.test=1, labels=se$label.fine, clusters=tcells$seurat_clusters)

new.cluster.ids <- c(test$labels) #Pull from clusters of singleR
names(new.cluster.ids) <- levels(tcells)
tcells <- RenameIdents(tcells, new.cluster.ids)

jpeg('./tcells/clustering//umap_labelled_singleRclusts_monaco_fine.jpg',height=800, width=1500)
DimPlot(tcells, reduction='umap', label=TRUE, label.size=6, pt.size=1.5)+ NoLegend()
dev.off()

Idents(tcells) <- 'seurat_clusters'
new.cluster.ids <- c('Th1 cells','CD8 T cells','TH17 and TRegs','Th2 cells','Effector CD8 T cells','Monocyte contaminant','NK cells','TH17 and TRegs','Monocyte contaminant')
names(new.cluster.ids) <- levels(tcells)
tcells <- RenameIdents(tcells, new.cluster.ids)
tcells[['Annotations']] <- Idents(object=tcells)
data$TCells <- tcells

jpeg('./tcells/clustering//umap_annotated.jpg',height=800, width=1500)
DimPlot(data$TCells, reduction='umap', group.by='Annotations',label=TRUE, label.size=6, pt.size=1.5)+ NoLegend()
dev.off()
#CD8A, FOXP3, GNLY, NKG7, CD4


Idents(monos) <- 'seurat_clusters'
se <- celldex::MonacoImmuneData() 
test <- SingleR(test=as.SingleCellExperiment(monos, assay='RNA'), ref=se, assay.type.test=1, labels=se$label.fine, clusters=monos$seurat_clusters)

#test <- SingleR(test=as.SingleCellExperiment(monos, assay='RNA'), ref=se, assay.type.test=1, labels=se$label.fine)

new.cluster.ids <- c(test$labels) #Pull from clusters of singleR
names(new.cluster.ids) <- levels(monos)
monos <- RenameIdents(monos, new.cluster.ids)

jpeg('./monos//clustering//umap_labelled_singleRclusts_monaco_fine.jpg',height=800, width=1500)
DimPlot(monos, reduction='umap', label=TRUE, label.size=6, pt.size=1.5)+ NoLegend()
dev.off()
Idents(monos) <- 'seurat_clusters'
new.cluster.ids <- c('MC1','MC2','MC3','MC4','MC5','MC6','MC7')
names(new.cluster.ids) <- levels(monos)
monos <- RenameIdents(monos, new.cluster.ids)
monos[['Annotations']] <- Idents(object=monos)
data$Monocytes <- monos


jpeg('./monos/clustering//umap_annotated.jpg',height=800, width=1500)
DimPlot(data$Monocytes, reduction='umap', group.by='Annotations',label=TRUE, label.size=6, pt.size=1.5)+ NoLegend()
dev.off()
```

```{r}
dir.create('pseudotime')
get_earliest_principal_node <- function(cds, origin){
                    cell_ids <- which(colData(cds)[, "seurat_clusters"] == origin)
                    
                    closest_vertex <-cds@principal_graph_aux[["UMAP"]]$pr_graph_cell_proj_closest_vertex
                    closest_vertex <- as.matrix(closest_vertex[colnames(cds), ])
                    root_pr_nodes <-igraph::V(principal_graph(cds)[["UMAP"]])$name[as.numeric(names(which.max(table(closest_vertex[cell_ids,]))))]
                    
                    root_pr_nodes
                }

mon <- as.cell_data_set(tcells)
mon <- cluster_cells(mon)
mon <- learn_graph(mon)

jpeg('./pseudotime/pseudotime_tcells.jpg',height=800, width=800)
mon0 <- order_cells(mon, root_pr_nodes=get_earliest_principal_node(mon, 0))
plot_cells(mon0, color_cells_by = "pseudotime",show_trajectory_graph = TRUE,cell_size = 0.75)
dev.off()

DefaultAssay(combined) <- 'SCT'
Idents(combined) <- 'seurat_clusters'
new.cluster.ids <- c('CD4 T','CD4 T','Monocytes','CD4 T','CD8 T','CD8 T','CD8 T','Dendritic cells','Monocytes','Erythroid cells','Dendritic cells','NK cells','NK cells','pDCs','B cells')
names(new.cluster.ids) <- levels(combined)
combined <- RenameIdents(combined, new.cluster.ids)
combined[['Annotations']] <- Idents(object=combined)
```

```{r TcellAnnotationPseudobulkPrep, warning=FALSE, message=FALSE, fig.align="center",results='hide'}
dir.create('./pseudobulk_annotated/')
dir.create('./pseudobulk_annotated/tcells/')
dir.create('./pseudobulk_annotated/tcells/counts')
tcells <- data$TCells
tcells$samp <- gsub("_","", as.character(tcells$samp))
Idents(tcells) <- 'samp'
tcells[['sample_id']] <- Idents(tcells)
Idents(tcells) <- 'Annotations'

Idents(tcells) <- 'group'
tcells[['group_id']] <- Idents(tcells)
Idents(tcells) <- 'Annotations'

counts <- tcells@assays$RNA@counts
metadata <- tcells@meta.data
metadata$cluster_id <- factor(tcells@active.ident)
sce <- SingleCellExperiment(assays=list(counts=counts), colData=metadata)
groups <- colData(sce, c("cluster_id", "sample_id"))

kids <- purrr::set_names(levels(sce$cluster_id))

nk <- length(kids)

sids <- purrr::set_names(levels(sce$sample_id))

ns <- length(sids)

n_cells <- as.numeric(table(sce$sample_id))

m <- match(sids, sce$sample_id)

ei <- data.frame(colData(sce)[m, ], 
    n_cells, row.names = NULL) %>% 
    dplyr::select(-"cluster_id")

groups <- colData(sce)[, c("cluster_id", "sample_id")]

pb <- aggregate.Matrix(t(counts(sce)), 
    groupings = groups, fun = "sum") 

splitf <- sapply(stringr::str_split(rownames(pb), 
                                    pattern = "_",  
                                    n = 2), 
                 `[`, 1)

pb <- split.data.frame(pb, 
                       factor(splitf)) %>%
        lapply(function(u) 
                set_colnames(t(u), 
                             stringr::str_extract(rownames(u), "(?<=_)[:alnum:]+")))

get_sample_ids <- function(x){
        pb[[x]] %>%
                colnames()
}

de_samples <- map(1:length(kids), get_sample_ids) %>%
        unlist()

samples_list <- map(1:length(kids), get_sample_ids)

get_cluster_ids <- function(x){
        rep(names(pb)[x], 
            each = length(samples_list[[x]]))
}

de_cluster_ids <- map(1:length(kids), get_cluster_ids) %>%
        unlist()


gg_df <- data.frame(cluster_id = de_cluster_ids,
                    sample_id = de_samples)

gg_df <- left_join(gg_df, ei[, c("sample_id", "group_id")]) 


metadata <- gg_df %>%
        dplyr::select(cluster_id, sample_id, group_id) 

clusters <- unique(metadata$cluster_id)
for(i in clusters){
    cluster_metadata <- metadata[which(metadata$cluster_id == paste(i)), ]
    rownames(cluster_metadata) <- cluster_metadata$sample_id
    counts <- pb[[paste(i)]]
    cluster_counts <- data.frame(counts[, which(colnames(counts) %in% rownames(cluster_metadata))])
    all(rownames(cluster_metadata) == colnames(cluster_counts))    
    dds <- DESeqDataSetFromMatrix(cluster_counts, 
                              colData = cluster_metadata, 
                              design = ~ group_id)
    dds <- DESeq(dds)
    vst <- DESeq2::varianceStabilizingTransformation(dds, blind=TRUE)
    mat <- assay(vst)

    compares <- list(
      list('CovidNormalTesting','control'),
      list('CovidPreRecovery','control'),
      list('CovidPersistent','control')
      
    )
    
    for(co in compares){
      run <- try({
        contrast <- c("group_id", co[[1]], co[[2]])
        res <- results(dds, 
            contrast = contrast,
            alpha = 0.05)
    
        res <- lfcShrink(dds, 
            contrast =  contrast,
            res=res,type="normal") 
        
        res <- res[order(res$padj),]
    
        res_tbl <- res %>%
            data.frame() %>%
            rownames_to_column(var="gene") %>%
            as_tibble()
        
        assign(paste0("cluster", i, "_", co[[1]], "_vs_", co[[2]]), res_tbl)
        jpeg(paste0("./pseudobulk_annotated/tcells/","volcano_", i, "_", co[[1]], "_vs_", co[[2]], ".jpg"), height=800, width=1000)
        p <- EnhancedVolcano(res_tbl,
                             lab=as.character(res_tbl$gene),
                             title=paste0(i, " ", co[[1]], " vs ", co[[2]], ".csv"),
                             x='log2FoldChange',
                             y='pvalue',
                             legendPosition = 'bottom',
                              legendLabels=c('Not sig.','Log (base 2) FC','nominal pvalue','nominal pvalue & Log (base 2) FC'),
                              pCutoff=0.05, 
                              FCcutoff=1
                             )
        print(p)
        dev.off()
        write.csv(res_tbl,
            paste0("./pseudobulk_annotated/tcells/","cluster", i, "_", co[[1]], "_vs_", co[[2]], "_all_genes.csv"),
            quote = FALSE, 
            row.names = FALSE)
        write.csv(as.data.frame(mat), file=paste0("./pseudobulk_annotated/tcells/counts/","cluster", i, "_vst_counts.csv"))  
        write.csv(counts(dds,normalized = TRUE), file=paste0("./pseudobulk_annotated/tcells/counts/", "cluster",i, "_dds_counts.csv")) 
       
        direct_compares[[paste0("tcell_cluster", i, "_", co[[1]], "_vs_", co[[2]])]] <- res_tbl
        mats[[paste0("tcell_cluster", i, "_", co[[1]], "_vs_", co[[2]])]] <- as.data.frame(mat)
      })
      if(inherits(run, 'try-error')) {
        next
      }
    }
}
```

```{r MonocyteAnnotationPseudobulkPrep, warning=FALSE, message=FALSE, fig.align="center",results='hide'}
monos <- data$Monocytes
dir.create('./pseudobulk_annotated/')
dir.create('./pseudobulk_annotated/monos/')
dir.create('./pseudobulk_annotated/monos/counts')
monos$samp <- gsub("_","", as.character(monos$samp))
Idents(monos) <- 'samp'
monos[['sample_id']] <- Idents(monos)
Idents(monos) <- 'Annotations'

Idents(monos) <- 'group'
monos[['group_id']] <- Idents(monos)
Idents(monos) <- 'Annotations'

counts <- monos@assays$RNA@counts
metadata <- monos@meta.data
metadata$cluster_id <- factor(monos@active.ident)
sce <- SingleCellExperiment(assays=list(counts=counts), colData=metadata)
groups <- colData(sce, c("cluster_id", "sample_id"))

kids <- purrr::set_names(levels(sce$cluster_id))

nk <- length(kids)

sids <- purrr::set_names(levels(sce$sample_id))

ns <- length(sids)

n_cells <- as.numeric(table(sce$sample_id))

m <- match(sids, sce$sample_id)

ei <- data.frame(colData(sce)[m, ], 
    n_cells, row.names = NULL) %>% 
    select(-"cluster_id")

groups <- colData(sce)[, c("cluster_id", "sample_id")]

pb <- aggregate.Matrix(t(counts(sce)), 
    groupings = groups, fun = "sum") 

splitf <- sapply(stringr::str_split(rownames(pb), 
                                    pattern = "_",  
                                    n = 2), 
                 `[`, 1)

pb <- split.data.frame(pb, 
                       factor(splitf)) %>%
        lapply(function(u) 
                set_colnames(t(u), 
                             stringr::str_extract(rownames(u), "(?<=_)[:alnum:]+")))

get_sample_ids <- function(x){
        pb[[x]] %>%
                colnames()
}

de_samples <- map(1:length(kids), get_sample_ids) %>%
        unlist()

samples_list <- map(1:length(kids), get_sample_ids)

get_cluster_ids <- function(x){
        rep(names(pb)[x], 
            each = length(samples_list[[x]]))
}

de_cluster_ids <- map(1:length(kids), get_cluster_ids) %>%
        unlist()


gg_df <- data.frame(cluster_id = de_cluster_ids,
                    sample_id = de_samples)

gg_df <- left_join(gg_df, ei[, c("sample_id", "group_id")]) 


metadata <- gg_df %>%
        dplyr::select(cluster_id, sample_id, group_id) 

clusters <- unique(metadata$cluster_id)
for(i in clusters){
    cluster_metadata <- metadata[which(metadata$cluster_id == paste(i)), ]
    rownames(cluster_metadata) <- cluster_metadata$sample_id
    counts <- pb[[paste(i)]]
    cluster_counts <- data.frame(counts[, which(colnames(counts) %in% rownames(cluster_metadata))])
    all(rownames(cluster_metadata) == colnames(cluster_counts))    
    dds <- DESeqDataSetFromMatrix(cluster_counts, 
                              colData = cluster_metadata, 
                              design = ~ group_id)
    dds <- DESeq(dds)
    vst <- DESeq2::varianceStabilizingTransformation(dds, blind=TRUE)
    mat <- assay(vst)

    compares <- list(
      list('CovidNormalTesting','control'),
      list('CovidPreRecovery','control'),
      list('CovidPersistent','control')
      
    )
    
    for(co in compares){
      run <- try({
        contrast <- c("group_id", co[[1]], co[[2]])
        res <- results(dds, 
            contrast = contrast,
            alpha = 0.05)
    
        res <- lfcShrink(dds, 
            contrast =  contrast,
            res=res,type="normal") 
        
        res <- res[order(res$padj),]
    
        res_tbl <- res %>%
            data.frame() %>%
            rownames_to_column(var="gene") %>%
            as_tibble()
        
        assign(paste0("cluster", i, "_", co[[1]], "_vs_", co[[2]]), res_tbl)
        jpeg(paste0("./pseudobulk_annotated/monos/","volcano_", i, "_", co[[1]], "_vs_", co[[2]], ".jpg"), height=800, width=1000)
        p <- EnhancedVolcano(res_tbl,
                             lab=as.character(res_tbl$gene),
                             title=paste0(i, " ", co[[1]], " vs ", co[[2]], ".csv"),
                             x='log2FoldChange',
                             y='pvalue',
                             legendPosition = 'bottom',
                              legendLabels=c('Not sig.','Log (base 2) FC','nominal pvalue','nominal pvalue & Log (base 2) FC'),
                              pCutoff=0.05, 
                              FCcutoff=1
                             )
        print(p)
        dev.off()
        write.csv(res_tbl,
            paste0("./pseudobulk_annotated/monos/","cluster", i, "_", co[[1]], "_vs_", co[[2]], "_all_genes.csv"),
            quote = FALSE, 
            row.names = FALSE)
        write.csv(as.data.frame(mat), file=paste0("./pseudobulk_annotated/monos/counts/","cluster", i, "_vst_counts.csv"))  
        write.csv(counts(dds,normalized = TRUE), file=paste0("./pseudobulk_annotated/monos/counts/", "cluster",i, "_dds_counts.csv")) 
        
        direct_compares[[paste0("mono_cluster", i, "_", co[[1]], "_vs_", co[[2]])]] <- res_tbl
        mats[[paste0("mono_cluster", i, "_", co[[1]], "_vs_", co[[2]])]] <- as.data.frame(mat)
      })
      if(inherits(run, 'try-error')) {
        next
      }
    }
}
```

```{r ModuleTesting, warning=FALSE, message=FALSE, fig.align="center",results='hide'}
mg <- c('ADORA3','ADRB2','CCR1','CXCR4','CX3CR1','GPR34','GPR183','C3AR1','C5AR1','FPR1','LPAR5','LPAR6','P2RY12','P2RY13','PTAFR','PTGER4','ADGRG1','IFIT3','IFITM3','ISG15','MRC1','CD83','EGR2','TNFSF18','CCL8','TFRC','CTSS')
test <- monos
DefaultAssay(test) <- 'SCT'
mgscore <- AddModuleScore(test, features=list(mg), name='Microglial Signature', ctrl=length(mg))
jpeg('./monocyte_microglial_signatures.jpg',height=1200, width=1200)
VlnPlot(mgscore, features='Microglial.Signature1', pt.size=0.5) + ggtitle("Microglial signature")+theme(axis.text.x=element_text(size=20), axis.text.y=element_text(size=20), legend.key.size=unit(1, 'cm'), legend.text=element_text(size=20), plot.title=element_text(size=24))
dev.off()



select_micro <- c('KCNQ1OT1','PDS5B','SYK','C1QC','APOE','BIN1','CTSL','LDHA','ANXA2','LMNA','FOXP1','ROCK1','TSPO','IMP3','ABI3','SLC3A2','GAS6','IL13RA1','CD63','LDHA','TIMP1','SPP1','MT-ND5','AKAP9','ARGLU1','DDOST','TWF2','RHEB')
jpeg('./vln_select_microglial_genes.jpg',height=2400, width=800)
VlnPlot(monos, features=select_micro,stack=TRUE, sort=TRUE, flip=TRUE) + theme(legend.position='none')
dev.off()

jpeg('./mg_score_umap.jpg',height=800, width=1500)
FeaturePlot(mgscore, reduction='umap',features = 'Microglial.Signature1', label=T, label.size=10,pt.size=1.5)+
            scale_colour_gradientn(colours = rev(brewer.pal(n = 11, name = "RdBu")))
dev.off()

```

```{r TCellPCATesting, warning=FALSE, message=FALSE, fig.align="center",results='hide'}
meta <- read.csv('./metadata.csv', header=T, row.names=4)
for(i in 1:length(mats)){
  metatemp <- meta[order(rownames(meta)),]
  p <- pca(mats[[i]][,order(names(mats[[i]]))], metadata = metatemp[rownames(metatemp) %in% colnames(mats[[i]]),])
  plot <- biplot(p, colby='subgroup',
               encircle=TRUE,
            legendPosition='top', legendLabSize=16, legendIconSize=8.0,
            pointSize=6,
            labSize=0)

  jpeg(paste0('./tcells/corrected_pca_nolabels_encircle',paste0(str_split(names(mats[i]),'_')[1][[1]][1],str_split(names(mats[i]),'_')[1][[1]][2]),'.jpg'), height=800, width=800)
  print(plot)
  dev.off()
}

```

```{r TCellPopulationShifts, warning=FALSE, message=FALSE, fig.align="center",results='hide'}
opt <- read.csv('./metadata.csv', header=T)
dir.create('PopShifts');
dir.create('PopShifts/tcells/')
Idents(data$TCells) <- 'seurat_clusters'
counts <- table(Idents(data$TCells), data$TCells$samp)
counts <- prop.table(counts, margin=2)*100

sigcounts <- data.frame(matrix(ncol=2,nrow=0)); colnames(sigcounts) <- c('cluster','pval')

temp <- as.data.frame(t(counts))
names(temp) <- c('sample_id','cluster_id','Freq')
temp$sample_id <- gsub("_","", as.character(temp$sample_id))

temp <-merge(temp, metadata %>% select(sample_id, group_id), by='sample_id') %>% distinct(sample_id, cluster_id,.keep_all=TRUE)



Idents(data$TCells) <- 'Annotations'
countsA <- table(Idents(data$TCells), data$TCells$samp)
countsA <- prop.table(countsA, margin=2)*100

sigcounts <- data.frame(matrix(ncol=2,nrow=0)); colnames(sigcounts) <- c('cluster','pval')

tempA <- as.data.frame(t(countsA))
names(tempA) <- c('sample_id','cluster_id','Freq')
tempA$sample_id <- gsub("_","", as.character(tempA$sample_id))

tempA <-merge(tempA, metadata %>% select(sample_id, group_id), by='sample_id') %>% distinct(sample_id, cluster_id,.keep_all=TRUE)


Idents(data$TCells) <- 'Annotations'
countsA_de <- as.data.frame(table(Idents(data$TCells), data$TCells$samp))
countsA_de <- spread(countsA_de, Var2, Freq)
countsA_de$rn <- gsub("\\+","", as.character(countsA_de$Var1))
countsA_de$rn <- gsub(" ","", as.character(countsA_de$rn))
countsA_de <- countsA_de %>% select(-c('Var1')) %>% column_to_rownames('rn')


metatemp <- opt
metatemp$name <- gsub("_","", as.character(metatemp$name))
metatemp <- metatemp %>% column_to_rownames('name')
metatemp <- metatemp[order(row.names(metatemp)),]
metatemp$subgroup <- factor(metatemp$subgroup)

dds <- DESeqDataSetFromMatrix(countsA_de, 
                              colData = metatemp, 
                              design = ~subgroup)
    

dds <- DESeq(dds)
compares <- list(
      list('CovidNormalTesting','control'),
      list('CovidPreRecovery','control'),
      list('CovidPersistent','control')
    )   

for(co in compares){
  run <- try({
    contrast <- c("subgroup", co[[1]], co[[2]])
    res <- results(dds, 
        contrast = contrast,
        alpha = 0.05)

    res <- lfcShrink(dds, 
        contrast =  contrast,
        res=res,type="normal") 
    
    res <- res[order(res$padj),]

    res_tbl <- res %>%
        data.frame() %>%
        rownames_to_column(var="gene") %>%
        as_tibble()
    
    write.csv(res_tbl,
            paste0("./PopShifts/tcells/", co[[1]], "_vs_", co[[2]], "_pairwise_popultion_shifts.csv"),
            quote = FALSE, 
            row.names = FALSE)
    
  })
  if(inherits(run, 'try-error')) {
    next
  }
}
for(i in unique(tempA$cluster_id)){
  jpeg(paste0('./PopShifts/tcells/cell_proportion_boxplot_',i,'.jpg'),height=800, width=800)
  p <- ggplot(tempA[tempA$cluster_id==paste(i),], aes(x=group_id, y=Freq)) + geom_boxplot(fill='#A4A4A4', color="black")+theme_classic() + labs(y="Cell proportion per sample (%)", x="Group") + theme(axis.text.x = element_text(angle = 45, hjust=1), text=element_text(size=20))
  print(p)
  dev.off()
}

write.csv(table(Idents(data$TCells),data$TCells$samp), file='.//tcell_counts_per_cluster.csv')
```

```{r MonocytePopulationShifts, warning=FALSE, message=FALSE, fig.align="center",results='hide'}
opt <- read.csv('./metadata.csv', header=T)
dir.create('PopShifts');
dir.create('PopShifts/monos/')
Idents(data$Monocytes) <- 'seurat_clusters'
counts <- table(Idents(data$Monocytes), data$Monocytes$samp)
counts <- prop.table(counts, margin=2)*100

sigcounts <- data.frame(matrix(ncol=2,nrow=0)); colnames(sigcounts) <- c('cluster','pval')

temp <- as.data.frame(t(counts))
names(temp) <- c('sample_id','cluster_id','Freq')
temp$sample_id <- gsub("_","", as.character(temp$sample_id))

temp <-merge(temp, metadata %>% select(sample_id, group_id), by='sample_id') %>% distinct(sample_id, cluster_id,.keep_all=TRUE)



Idents(data$Monocytes) <- 'Annotations'
countsA <- table(Idents(data$Monocytes), data$Monocytes$samp)
countsA <- prop.table(countsA, margin=2)*100

sigcounts <- data.frame(matrix(ncol=2,nrow=0)); colnames(sigcounts) <- c('cluster','pval')

tempA <- as.data.frame(t(countsA))
names(tempA) <- c('sample_id','cluster_id','Freq')
tempA$sample_id <- gsub("_","", as.character(tempA$sample_id))

tempA <-merge(tempA, metadata %>% select(sample_id, group_id), by='sample_id') %>% distinct(sample_id, cluster_id,.keep_all=TRUE)


Idents(data$Monocytes) <- 'Annotations'
countsA_de <- as.data.frame(table(Idents(data$Monocytes), data$Monocytes$samp))
countsA_de <- spread(countsA_de, Var2, Freq)
countsA_de$rn <- gsub("\\+","", as.character(countsA_de$Var1))
countsA_de$rn <- gsub(" ","", as.character(countsA_de$rn))
countsA_de <- countsA_de %>% select(-c('Var1')) %>% column_to_rownames('rn')


metatemp <- opt
metatemp$name <- gsub("_","", as.character(metatemp$name))
metatemp <- metatemp %>% column_to_rownames('name')
metatemp <- metatemp[order(row.names(metatemp)),]
metatemp$subgroup <- factor(metatemp$subgroup)

dds <- DESeqDataSetFromMatrix(countsA_de, 
                              colData = metatemp, 
                              design = ~subgroup)
    

dds <- DESeq(dds)
compares <- list(
      list('CovidNormalTesting','control'),
      list('CovidPreRecovery','control'),
      list('CovidPersistent','control')
    )   

for(co in compares){
  run <- try({
    contrast <- c("subgroup", co[[1]], co[[2]])
    res <- results(dds, 
        contrast = contrast,
        alpha = 0.05)

    res <- lfcShrink(dds, 
        contrast =  contrast,
        res=res,type="normal") 
    
    res <- res[order(res$padj),]

    res_tbl <- res %>%
        data.frame() %>%
        rownames_to_column(var="gene") %>%
        as_tibble()
    
    write.csv(res_tbl,
            paste0("./PopShifts/monos/", co[[1]], "_vs_", co[[2]], "_pairwise_popultion_shifts.csv"),
            quote = FALSE, 
            row.names = FALSE)
    
  })
  if(inherits(run, 'try-error')) {
    next
  }
}
for(i in unique(tempA$cluster_id)){
  jpeg(paste0('./PopShifts/monos/cell_proportion_boxplot_',i,'.jpg'),height=800, width=800)
  p <- ggplot(tempA[tempA$cluster_id==paste(i),], aes(x=group_id, y=Freq)) + geom_boxplot(fill='#A4A4A4', color="black")+theme_classic() + labs(y="Cell proportion per sample (%)", x="Group") + theme(axis.text.x = element_text(angle = 45, hjust=1), text=element_text(size=20))
  print(p)
  dev.off()
}

write.csv(table(Idents(data$Monocytes),data$Monocytes$samp), file='.//monocytes_counts_per_cluster.csv')
```

```{r TCellGSEATesting, warning=FALSE, message=FALSE, fig.align="center",results='hide'}
library(clusterProfiler)
library(msigdbr)
library(stringr)
library(enrichplot)
dir.create('./GSEA')
dir.create('./GSEA/tcells')
dir.create('./GSEA/tcells/MSigDB_hallmark')
dir.create('./GSEA/tcells/MSigDB_C2')
dir.create('./GSEA/tcells/KEGG')
dir.create('./GSEA/tcells/GO')
organism='org.Hs.eg.db'
korg='hsa'
msig_org='Homo sapiens'

compares <- list(
  list('CD8 CovidPreRecovery vs control',`clusterCD8 T cells_CovidPreRecovery_vs_control`),
  list('CD8 Persistent vs Control',`clusterCD8 T cells_CovidPersistent_vs_control`),
  list('CD8 Covid Persistent vs Covid PreRecovery',`clusterCD8 T cells_CovidPersistent_vs_CovidPreRecovery`),
  list('Effector CD8 CovidPreRecovery vs control',`clusterEffector CD8 T cells_CovidPreRecovery_vs_control`),
  list('Effector CD8 Persistent vs Control',`clusterEffector CD8 T cells_CovidPersistent_vs_control`),
  list('Effector CD8 Covid Persistent vs Covid PreRecovery',`clusterEffector CD8 T cells_CovidPersistent_vs_CovidPreRecovery`),
  list('NK CovidPreRecovery vs control',`clusterNK cells_CovidPreRecovery_vs_control`),
  list('NK Persistent vs Control',`clusterNK cells_CovidPersistent_vs_control`),
  list('NK Covid Persistent vs Covid PreRecovery',`clusterNK cells_CovidPersistent_vs_CovidPreRecovery`),
  list('TH1 CovidPreRecovery vs control',`clusterTh1 cells_CovidPreRecovery_vs_control`),
  list('TH1 Persistent vs Control',`clusterTh1 cells_CovidPersistent_vs_control`),
  list('TH1 Covid Persistent vs Covid PreRecovery',`clusterTh1 cells_CovidPersistent_vs_CovidPreRecovery`),
  list('TH17 and Treg CovidPreRecovery vs control',`clusterTH17 and TRegs_CovidPreRecovery_vs_control`),
  list('TH17 and Treg Persistent vs Control',`clusterTH17 and TRegs_CovidPersistent_vs_control`),
  list('TH17 and Treg Covid Persistent vs Covid PreRecovery',`clusterTH17 and TRegs_CovidPersistent_vs_CovidPreRecovery`),
  list('TH2 CovidPreRecovery vs control',`clusterTh2 cells_CovidPreRecovery_vs_control`),
  list('TH2 Persistent vs Control',`clusterTh2 cells_CovidPersistent_vs_control`),
  list('TH2 Covid Persistent vs Covid PreRecovery',`clusterTh2 cells_CovidPersistent_vs_CovidPreRecovery`)
)

#MSigDB hallmark
m_t2g <- msigdbr(species = msig_org, category = "H") %>% 
  dplyr::select(gs_name, entrez_gene)

for (DataSet in compares){
  DataIn <- as.data.frame(DataSet[2])
  genelist <- DataIn[,2]
  names(genelist) <- as.character(DataIn[,1])
  genelist <- sort(genelist, decreasing=T)
  ids <- bitr(names(genelist), fromType='SYMBOL',toType='ENTREZID',OrgDb=organism)
  
  tab <- DataIn %>% dplyr::select(gene, stat)
  names(tab) <- c('SYMBOL','STAT')
  merged <- merge(tab, ids, by='SYMBOL')
  
  genelist2 <- merged[,2]
  names(genelist2) <- as.character(merged[,3])
  genelist2[genelist2==0] <- NA
  genelist2 <- sort(na.omit(genelist2), decreasing=T)
  
  run <- try({
    gse <- GSEA(genelist2, TERM2GENE=m_t2g, nPerm=1000, pvalueCutoff=1, minGSSize=3, maxGSSize=800)
    plot <- dotplot(gse, showCategory=10, font.size=10, split=".sign") + facet_grid(.~.sign) + scale_y_discrete(labels=function(x) str_wrap(x, width=40))
    jpeg(paste('./GSEA/tcells/MSigDB_hallmark/',DataSet[1],'_msigdb_H_GSEA.jpg',sep=''), height=800, width=1400)
    print(plot)
    dev.off()
      
    write.csv(gse, file=paste('./GSEA/tcells/MSigDB_hallmark/',DataSet[1],'_msigdb_H_GSEA.csv',sep=''))
  })
  if(inherits(run, 'try-error')){
    next
  }
}

#MSigDB C2
m_t2g <- msigdbr(species = msig_org, category = "C2") %>% 
  dplyr::select(gs_name, entrez_gene)

for (DataSet in compares){
  DataIn <- as.data.frame(DataSet[2])
  genelist <- DataIn[,2]
  names(genelist) <- as.character(DataIn[,1])
  genelist <- sort(genelist, decreasing=T)
  ids <- bitr(names(genelist), fromType='SYMBOL',toType='ENTREZID',OrgDb=organism)
  
  tab <- DataIn %>% dplyr::select(gene, stat)
  names(tab) <- c('SYMBOL','STAT')
  merged <- merge(tab, ids, by='SYMBOL')
  
  genelist2 <- merged[,2]
  names(genelist2) <- as.character(merged[,3])
  genelist2[genelist2==0] <- NA
  genelist2 <- sort(na.omit(genelist2), decreasing=T)
  
  run <- try({
    gse <- GSEA(genelist2, TERM2GENE=m_t2g, nPerm=1000, pvalueCutoff=1, minGSSize=3, maxGSSize=800)
    plot <- dotplot(gse, showCategory=10, font.size=10,split=".sign") + facet_grid(.~.sign) + scale_y_discrete(labels=function(x) str_wrap(x, width=40))
    jpeg(paste('./GSEA/tcells/MSigDB_C2/',DataSet[1],'_msigdb_C2_GSEA.jpg',sep=''), height=800, width=1400)
    print(plot)
    dev.off()
      
    write.csv(gse, file=paste('./GSEA/tcells/MSigDB_C2/',DataSet[1],'_msigdb_C2_GSEA.csv',sep=''))
  })
  if(inherits(run, 'try-error')){
    next
  }
}

#KEGG
for (DataSet in compares){
  DataIn <- as.data.frame(DataSet[2])
  genelist <- DataIn[,2]
  names(genelist) <- as.character(DataIn[,1])
  genelist <- sort(genelist, decreasing=T)
  ids <- bitr(names(genelist), fromType='SYMBOL',toType='ENTREZID',OrgDb=organism)
  
  tab <- DataIn %>% dplyr::select(gene, stat)
  names(tab) <- c('SYMBOL','STAT')
  merged <- merge(tab, ids, by='SYMBOL')
  
  genelist2 <- merged[,2]
  names(genelist2) <- as.character(merged[,3])
  genelist2[genelist2==0] <- NA
  genelist2 <- sort(na.omit(genelist2), decreasing=T)
  
  run <- try({
    gse <- gseKEGG(genelist2, nPerm=1000, pvalueCutoff=1, minGSSize=3, maxGSSize=800, organism=korg)
    plot <- dotplot(gse, showCategory=10, font.size=10,split=".sign") + facet_grid(.~.sign) + scale_y_discrete(labels=function(x) str_wrap(x, width=40))
    jpeg(paste('./GSEA/tcells/KEGG/',DataSet[1],'_KEGG_GSEA.jpg',sep=''), height=800, width=1400)
    print(plot)
    dev.off()
      
    write.csv(gse, file=paste('./GSEA/tcells/KEGG/',DataSet[1],'_KEGG_GSEA.csv',sep=''))
  })
  if(inherits(run, 'try-error')){
    next
  }
}

#GO
for (DataSet in compares){
  DataIn <- as.data.frame(DataSet[2])
  genelist <- DataIn[,2]
  names(genelist) <- as.character(DataIn[,1])
  genelist <- sort(genelist, decreasing=T)
  ids <- bitr(names(genelist), fromType='SYMBOL',toType='ENTREZID',OrgDb=organism)
  
  tab <- DataIn %>% dplyr::select(gene, stat)
  names(tab) <- c('SYMBOL','STAT')
  merged <- merge(tab, ids, by='SYMBOL')
  
  genelist2 <- merged[,2]
  names(genelist2) <- as.character(merged[,3])
  genelist2[genelist2==0] <- NA
  genelist2 <- sort(na.omit(genelist2), decreasing=T)
  
  run <- try({
    gse <- gseGO(genelist2, ont='BP', nPerm=1000, pvalueCutoff=1, minGSSize=3, maxGSSize=800, OrgDb=organism)
    plot <- dotplot(gse, showCategory=10, font.size=10,split=".sign") + facet_grid(.~.sign) + scale_y_discrete(labels=function(x) str_wrap(x, width=40))
    jpeg(paste('./GSEA/tcells/GO/',DataSet[1],'_GO_GSEA.jpg',sep=''), height=800, width=1400)
    print(plot)
    dev.off()
      
    write.csv(gse, file=paste('./GSEA/tcells/GO/',DataSet[1],'_GO_GSEA.csv',sep=''))
  })
  if(inherits(run, 'try-error')){
    next
  }
}
```

```{r MonocyteGSEATesting, warning=FALSE, message=FALSE, fig.align="center",results='hide'}
library(clusterProfiler)
library(msigdbr)
library(stringr)
library(enrichplot)
dir.create('./GSEA')
dir.create('./GSEA/monos')
dir.create('./GSEA/monos/MSigDB_hallmark')
dir.create('./GSEA/monos/MSigDB_C2')
dir.create('./GSEA/monos/KEGG')
dir.create('./GSEA/monos/GO')
organism='org.Hs.eg.db'
korg='hsa'
msig_org='Homo sapiens'

gselist <- list()

compares <- list(
  list('MC1 CovidPreRecovery vs control',clusterMC1_CovidPreRecovery_vs_control),
  list('MC1 Covid Persistent vs control',clusterMC1_CovidPersistent_vs_control),
  list('MC1 Covid Persistent vs Covid PreRecovery',clusterMC1_CovidPersistent_vs_CovidPreRecovery),
  
  list('MC2 CovidPreRecovery vs control',clusterMC2_CovidPreRecovery_vs_control),
  list('MC2 Covid Persistent vs control',clusterMC2_CovidPersistent_vs_control),
  list('MC2 Covid Persistent vs Covid PreRecovery',clusterMC2_CovidPersistent_vs_CovidPreRecovery),
  
  list('MC3 CovidPreRecovery vs control',clusterMC3_CovidPreRecovery_vs_control),
  list('MC3 Covid Persistent vs control',clusterMC3_CovidPersistent_vs_control),
  list('MC3 Covid Persistent vs Covid PreRecovery',clusterMC3_CovidPersistent_vs_CovidPreRecovery),
  
  list('MC4 CovidPreRecovery vs control',clusterMC4_CovidPreRecovery_vs_control),
  list('MC4 Covid Persistent vs control',clusterMC4_CovidPersistent_vs_control),
  list('MC4 Covid Persistent vs Covid PreRecovery',clusterMC4_CovidPersistent_vs_CovidPreRecovery),
  
  list('MC5 CovidPreRecovery vs control',clusterMC5_CovidPreRecovery_vs_control),
  list('MC5 Covid Persistent vs control',clusterMC5_CovidPersistent_vs_control),
  list('MC5 Covid Persistent vs Covid PreRecovery',clusterMC5_CovidPersistent_vs_CovidPreRecovery),
  
  list('MC6 CovidPreRecovery vs control',clusterMC6_CovidPreRecovery_vs_control),
  list('MC6 Covid Persistent vs control',clusterMC6_CovidPersistent_vs_control),
  list('MC6 Covid Persistent vs Covid PreRecovery',clusterMC6_CovidPersistent_vs_CovidPreRecovery),
  
  list('MC7 CovidPreRecovery vs control',clusterMC7_CovidPreRecovery_vs_control),
  list('MC7 Covid Persistent vs control',clusterMC7_CovidPersistent_vs_control),
  list('MC7 Covid Persistent vs Covid PreRecovery',clusterMC7_CovidPersistent_vs_CovidPreRecovery)
)

#MSigDB hallmark
m_t2g <- msigdbr(species = msig_org, category = "H") %>% 
  dplyr::select(gs_name, entrez_gene)

for (DataSet in compares){
  DataIn <- as.data.frame(DataSet[2])
  genelist <- DataIn[,2]
  names(genelist) <- as.character(DataIn[,1])
  genelist <- sort(genelist, decreasing=T)
  ids <- bitr(names(genelist), fromType='SYMBOL',toType='ENTREZID',OrgDb=organism)
  
  tab <- DataIn %>% dplyr::select(gene, stat)
  names(tab) <- c('SYMBOL','STAT')
  merged <- merge(tab, ids, by='SYMBOL')
  
  genelist2 <- merged[,2]
  names(genelist2) <- as.character(merged[,3])
  genelist2 <- sort(na.omit(genelist2), decreasing=T)
  
  run <- try({
    gse <- GSEA(genelist2, TERM2GENE=m_t2g, nPerm=1000, pvalueCutoff=1, minGSSize=3, maxGSSize=800)
    plot <- dotplot(gse, showCategory=10, font.size=10, split=".sign") + facet_grid(.~.sign) + scale_y_discrete(labels=function(x) str_wrap(x, width=40))
    jpeg(paste('./GSEA/monos/MSigDB_hallmark/',DataSet[1],'_msigdb_H_GSEA.jpg',sep=''), height=800, width=1400)
    print(plot)
    dev.off()
      
    gselist[[paste0(DataSet[[1]],'_H')]] <- gse
    write.csv(gse, file=paste('./GSEA/monos/MSigDB_hallmark/',DataSet[1],'_msigdb_H_GSEA.csv',sep=''))
  })
  if(inherits(run, 'try-error')){
    next
  }
}

#MSigDB C2
m_t2g <- msigdbr(species = msig_org, category = "C2") %>% 
  dplyr::select(gs_name, entrez_gene)

for (DataSet in compares){
  DataIn <- as.data.frame(DataSet[2])
  genelist <- DataIn[,2]
  names(genelist) <- as.character(DataIn[,1])
  genelist <- sort(genelist, decreasing=T)
  ids <- bitr(names(genelist), fromType='SYMBOL',toType='ENTREZID',OrgDb=organism)
  
  tab <- DataIn %>% dplyr::select(gene, stat)
  names(tab) <- c('SYMBOL','STAT')
  merged <- merge(tab, ids, by='SYMBOL')
  
  genelist2 <- merged[,2]
  names(genelist2) <- as.character(merged[,3])
  genelist2[genelist2==0] <- NA
  genelist2 <- sort(na.omit(genelist2), decreasing=T)
  
  run <- try({
    gse <- GSEA(genelist2, TERM2GENE=m_t2g, nPerm=1000, pvalueCutoff=1, minGSSize=3, maxGSSize=800)
    plot <- dotplot(gse, showCategory=10, font.size=10,split=".sign") + facet_grid(.~.sign) + scale_y_discrete(labels=function(x) str_wrap(x, width=40))
    jpeg(paste('./GSEA/monos/MSigDB_C2/',DataSet[1],'_msigdb_C2_GSEA.jpg',sep=''), height=800, width=1400)
    print(plot)
    dev.off()
      
    gselist[[paste0(DataSet[[1]],'_C2')]] <- gse
    write.csv(gse, file=paste('./GSEA/monos/MSigDB_C2/',DataSet[1],'_msigdb_C2_GSEA.csv',sep=''))
  })
  if(inherits(run, 'try-error')){
    next
  }
}

#KEGG
for (DataSet in compares){
  DataIn <- as.data.frame(DataSet[2])
  genelist <- DataIn[,2]
  names(genelist) <- as.character(DataIn[,1])
  genelist <- sort(genelist, decreasing=T)
  ids <- bitr(names(genelist), fromType='SYMBOL',toType='ENTREZID',OrgDb=organism)
  
  tab <- DataIn %>% dplyr::select(gene, stat)
  names(tab) <- c('SYMBOL','STAT')
  merged <- merge(tab, ids, by='SYMBOL')
  
  genelist2 <- merged[,2]
  names(genelist2) <- as.character(merged[,3])
  genelist2 <- sort(na.omit(genelist2), decreasing=T)
  
  run <- try({
    gse <- gseKEGG(genelist2, nPerm=1000, pvalueCutoff=1, minGSSize=3, maxGSSize=800, organism=korg)
    plot <- dotplot(gse, showCategory=10, font.size=10,split=".sign") + facet_grid(.~.sign) + scale_y_discrete(labels=function(x) str_wrap(x, width=40))
    jpeg(paste('./GSEA/monos/KEGG/',DataSet[1],'_KEGG_GSEA.jpg',sep=''), height=800, width=1400)
    print(plot)
    dev.off()
      
    gselist[[paste0(DataSet[[1]],'_KEGG')]] <- gse
    write.csv(gse, file=paste('./GSEA/monos/KEGG/',DataSet[1],'_KEGG_GSEA.csv',sep=''))
  })
  if(inherits(run, 'try-error')){
    next
  }
}

#GO
for (DataSet in compares){
  DataIn <- as.data.frame(DataSet[2])
  genelist <- DataIn[,2]
  names(genelist) <- as.character(DataIn[,1])
  genelist <- sort(genelist, decreasing=T)
  ids <- bitr(names(genelist), fromType='SYMBOL',toType='ENTREZID',OrgDb=organism)
  
  tab <- DataIn %>% dplyr::select(gene, stat)
  names(tab) <- c('SYMBOL','STAT')
  merged <- merge(tab, ids, by='SYMBOL')
  
  genelist2 <- merged[,2]
  names(genelist2) <- as.character(merged[,3])
  genelist2 <- sort(na.omit(genelist2), decreasing=T)
  
  run <- try({
    gse <- gseGO(genelist2, ont='BP', nPerm=1000, pvalueCutoff=1, minGSSize=3, maxGSSize=800, OrgDb=organism)
    plot <- dotplot(gse, showCategory=10, font.size=10,split=".sign") + facet_grid(.~.sign) + scale_y_discrete(labels=function(x) str_wrap(x, width=40))
    jpeg(paste('./GSEA/monos/GO/',DataSet[1],'_GO_GSEA.jpg',sep=''), height=800, width=1400)
    print(plot)
    dev.off()
      
    gselist[[paste0(DataSet[[1]],'_GO')]] <- gse
    write.csv(gse, file=paste('./GSEA/monos/GO/',DataSet[1],'_GO_GSEA.csv',sep=''))
  })
  if(inherits(run, 'try-error')){
    next
  }
}
```

```{r SplitByGroupingPlots, warning=FALSE, message=FALSE, fig.align="center",results='hide'}
monossub <- subset(monos, subset=subgroup==c('control','CovidNormalTesting','CovidPersistent','CovidPreRecovery'))

jpeg('Impaired_vs_Control.jpg',height=1600,width=800)
title=ggdraw() + draw_label('Select genes from Impaired vs Control')
row1=cowplot::plot_grid(ggdraw() + draw_label('NK cells: LZTFL1'), ggdraw() + draw_label('NK cells: NISCH'), ncol=2, rel_widths=c(4,4))
row2=cowplot::plot_grid(VlnPlot(tcells, features=c('LZTFL1'), idents = 'NK cells',group.by='group') + theme(plot.title = element_text(size=0),axis.title.x=element_text(size=0)) + NoLegend(),VlnPlot(tcells, features=c('NISCH'), idents = 'NK cells',group.by='group') + theme(plot.title = element_text(size=0),axis.title.x=element_text(size=0)) + NoLegend(), ncol=2, rel_widths=c(4,4))

row3=cowplot::plot_grid(ggdraw() + draw_label('MC6 cells: GABARAPL1'), ggdraw() + draw_label('MC7 cells: TCF4'), ncol=2, rel_widths=c(4,4))
row4=cowplot::plot_grid(VlnPlot(monos, features=c('GABARAPL1'), idents = 'MC6',group.by='group') + theme(plot.title = element_text(size=0),axis.title.x=element_text(size=0)) + NoLegend(),VlnPlot(monos, features=c('TCF4'), idents = 'MC7',group.by='group') + theme(plot.title = element_text(size=0),axis.title.x=element_text(size=0)) + NoLegend(), ncol=2, rel_widths=c(4,4))

row5=cowplot::plot_grid(ggdraw() + draw_label('MC7 cells: ID2'), ggdraw() + draw_label('MC7 cells: GNAQ'), ncol=2, rel_widths=c(4,4))
row6=cowplot::plot_grid(VlnPlot(monos, features=c('ID2'), idents = 'MC7',group.by='group') + theme(plot.title = element_text(size=0),axis.title.x=element_text(size=0)) + NoLegend(),VlnPlot(monos, features=c('GNAQ'), idents = 'MC7',group.by='group') + theme(plot.title = element_text(size=0),axis.title.x=element_text(size=0)) + NoLegend(), ncol=2, rel_widths=c(4,4))

row7=cowplot::plot_grid(ggdraw() + draw_label('MC7 cells: IER5'), ggdraw() + draw_label('MC7 cells: PMAIP1'), ncol=2, rel_widths=c(4,4))
row8=cowplot::plot_grid(VlnPlot(monos, features=c('IER5'), idents = 'MC7',group.by='group') + theme(plot.title = element_text(size=0),axis.title.x=element_text(size=0)) + NoLegend(),VlnPlot(monos, features=c('PMAIP1'), idents = 'MC7',group.by='group') + theme(plot.title = element_text(size=0),axis.title.x=element_text(size=0)) + NoLegend(), ncol=2, rel_widths=c(4,4))

row9=cowplot::plot_grid(ggdraw() + draw_label('MC7 cells: NAGK'), ggdraw() + draw_label('MC7 cells: ARRDC3'), ncol=2, rel_widths=c(4,4))
row10=cowplot::plot_grid(VlnPlot(monos, features=c('NAGK'), idents = 'MC7',group.by='group') + theme(plot.title = element_text(size=0),axis.title.x=element_text(size=0)) + NoLegend(),VlnPlot(monos, features=c('ARRDC3'), idents = 'MC7',group.by='group') + theme(plot.title = element_text(size=0),axis.title.x=element_text(size=0)) + NoLegend(), ncol=2, rel_widths=c(4,4))

cowplot::plot_grid(title, row1, row2,row3,row4,row5,row6,row7,row8,row9,row10, ncol=1, rel_heights=c(0.3,0.25,5,0.25,5,0.25,5,0.25,5,0.25,5))
dev.off()

############################################

jpeg('Persistent_vs_Recovered.jpg',height=1600,width=1200)
title=ggdraw() + draw_label('Select genes from Persistent vs Recovered')
row1=cowplot::plot_grid(ggdraw() + draw_label('MC1 cells: CD1C'), ggdraw() + draw_label('MC1 cells: SMARCD3'), ggdraw() + draw_label('MC1 cells: JUP'),ncol=3, rel_widths=c(4,4,4))
row2=cowplot::plot_grid(VlnPlot(monossub, features=c('CD1C'), idents = 'MC1',group.by='subgroup') + theme(plot.title = element_text(size=0),axis.title.x=element_text(size=0)) + NoLegend(),VlnPlot(monossub, features=c('SMARCD3'), idents = 'MC1',group.by='subgroup') + theme(plot.title = element_text(size=0),axis.title.x=element_text(size=0)) + NoLegend(), VlnPlot(monossub, features=c('JUP'), idents = 'MC1',group.by='subgroup') + theme(plot.title = element_text(size=0),axis.title.x=element_text(size=0)) + NoLegend(),ncol=3, rel_widths=c(4,4,4))

row3=cowplot::plot_grid(ggdraw() + draw_label('MC1 cells: FCN1'), ggdraw() + draw_label('MC2 cells: HBA2'), ggdraw() + draw_label('MC2 cells: HAUS2'),ncol=3, rel_widths=c(4,4,4))
row4=cowplot::plot_grid(VlnPlot(monossub, features=c('FCN1'), idents = 'MC1',group.by='subgroup') + theme(plot.title = element_text(size=0),axis.title.x=element_text(size=0)) + NoLegend(),VlnPlot(monossub, features=c('HBA2'), idents = 'MC2',group.by='subgroup') + theme(plot.title = element_text(size=0),axis.title.x=element_text(size=0)) + NoLegend(), VlnPlot(monossub, features=c('HAUS2'), idents = 'MC2',group.by='subgroup') + theme(plot.title = element_text(size=0),axis.title.x=element_text(size=0)) + NoLegend(),ncol=3, rel_widths=c(4,4,4))

row5=cowplot::plot_grid(ggdraw() + draw_label('MC2 cells: FN1'), ggdraw() + draw_label('MC3 cells: TAP1'), ggdraw() + draw_label('MC3 cells: JUP'),ncol=3, rel_widths=c(4,4,4))
row6=cowplot::plot_grid(VlnPlot(monossub, features=c('FN1'), idents = 'MC2',group.by='subgroup') + theme(plot.title = element_text(size=0),axis.title.x=element_text(size=0)) + NoLegend(),VlnPlot(monossub, features=c('TAP1'), idents = 'MC3',group.by='subgroup') + theme(plot.title = element_text(size=0),axis.title.x=element_text(size=0)) + NoLegend(), VlnPlot(monossub, features=c('JUP'), idents = 'MC3',group.by='subgroup') + theme(plot.title = element_text(size=0),axis.title.x=element_text(size=0)) + NoLegend(),ncol=3, rel_widths=c(4,4,4))

row7=cowplot::plot_grid(ggdraw() + draw_label('MC4 cells: SUMO4'), ggdraw() + draw_label('MC5 cells: HBA2'), ggdraw() + draw_label('MC6 cells: HIST1H1D'),ncol=3, rel_widths=c(4,4,4))
row8=cowplot::plot_grid(VlnPlot(monossub, features=c('SUMO4'), idents = 'MC4',group.by='subgroup') + theme(plot.title = element_text(size=0),axis.title.x=element_text(size=0)) + NoLegend(),VlnPlot(monossub, features=c('HBA2'), idents = 'MC5',group.by='subgroup') + theme(plot.title = element_text(size=0),axis.title.x=element_text(size=0)) + NoLegend(), VlnPlot(monossub, features=c('HIST1H1D'), idents = 'MC6',group.by='subgroup') + theme(plot.title = element_text(size=0),axis.title.x=element_text(size=0)) + NoLegend(),ncol=3, rel_widths=c(4,4,4))

cowplot::plot_grid(title, row1, row2,row3,row4,row5,row6,row7,row8, ncol=1, rel_heights=c(0.3,0.25,5,0.25,5,0.25,5,0.25,5))
dev.off()

mc1gse <- gselist$`MC1 Cognitively Impaired vs Control_C2` %>% filter(., p.adjust < 0.05)
mc1gse <- pairwise_termsim(mc1gse)

mc2gse <- gselist$`MC2 Cognitively Impaired vs Control_C2` %>% filter(., p.adjust < 0.05)
mc2gse <- pairwise_termsim(mc2gse)

mc4gse <- gselist$`MC4 Cognitively Impaired vs Control_C2` %>% filter(., p.adjust < 0.05)
mc4gse <- pairwise_termsim(mc4gse)

mc7gse <- gselist$`MC7 Cognitively Impaired vs Control_C2` %>% filter(., p.adjust < 0.05)
mc7gse <- pairwise_termsim(mc7gse)

title=ggdraw() + draw_label('Top 50 C2 Pathways')
row1=cowplot::plot_grid(ggdraw() + draw_label('MC1 CI vs Control'), ggdraw() + draw_label('MC2 CI vs Control'),ncol=2, rel_widths=c(4,4))
row2=cowplot::plot_grid(emapplot(mc1gse, showCategory = 50, color='pvalue'),emapplot(mc2gse, showCategory = 50, color='pvalue'),ncol=2, rel_widths=c(4,4))
row3=cowplot::plot_grid(ggdraw() + draw_label('MC4 CI vs Control'), ggdraw() + draw_label('MC7 CI vs Control'),ncol=2, rel_widths=c(4,4))
row4=cowplot::plot_grid(emapplot(mc4gse, showCategory = 50, color='pvalue'),emapplot(mc7gse, showCategory = 50, color='pvalue'),ncol=2, rel_widths=c(4,4))


jpeg('./test.jpg', height=2400, width=2400)
cowplot::plot_grid(title, row1, row2,row3,row4, ncol=1, rel_heights=c(0.3,0.25,5,0.25,5))
dev.off()

pdf('./test.pdf', height=50, width=50)
cowplot::plot_grid(title, row1, row2,row3,row4, ncol=1, rel_heights=c(0.3,0.25,5,0.25,5))
dev.off()

###################################

mc1gse <- gselist$`MC1 Cognitively Impaired vs Control_C2` %>% filter(., grepl('BLANCO_MELO',ID))
mc1gse <- pairwise_termsim(mc1gse)

mc2gse <- gselist$`MC2 Cognitively Impaired vs Control_C2` %>% filter(., grepl('BLANCO_MELO',ID))
mc2gse <- pairwise_termsim(mc2gse)

mc4gse <- gselist$`MC4 Cognitively Impaired vs Control_C2` %>% filter(., grepl('BLANCO_MELO',ID))
mc4gse <- pairwise_termsim(mc4gse)

mc7gse <- gselist$`MC7 Cognitively Impaired vs Control_C2` %>% filter(., grepl('BLANCO_MELO',ID))
mc7gse <- pairwise_termsim(mc7gse)

title=ggdraw() + draw_label('Top Blanco Melo Pathways')
row1=cowplot::plot_grid(ggdraw() + draw_label('MC1 CI vs Control'), ggdraw() + draw_label('MC2 CI vs Control'),ncol=2, rel_widths=c(4,4))
row2=cowplot::plot_grid(emapplot(mc1gse, showCategory = 50, color='pvalue'),emapplot(mc2gse, showCategory = 50, color='pvalue'),ncol=2, rel_widths=c(4,4))
row3=cowplot::plot_grid(ggdraw() + draw_label('MC4 CI vs Control'), ggdraw() + draw_label('MC7 CI vs Control'),ncol=2, rel_widths=c(4,4))
row4=cowplot::plot_grid(emapplot(mc4gse, showCategory = 50, color='pvalue'),emapplot(mc7gse, showCategory = 50, color='pvalue'),ncol=2, rel_widths=c(4,4))
pdf('./test_blanco_melo.pdf', height=50, width=50)
cowplot::plot_grid(title, row1, row2,row3,row4, ncol=1, rel_heights=c(0.3,0.25,5,0.25,5))
dev.off()

save.image('data.RData')
```